### ISSUE TITLE
`odo dev`

### ISSUE THREAD

**Issue by serenamarie125** [redhat-developer/odo#5299]
description by @kadel:

# `odo dev`

## User Story
- As an odo user I want to be able to quickly iterate on my application running on the cluster so that 
  - I can run the application in the environment where it will be deployed
  - I can easily consume in-cluster services
  - I can quickly see changes that I made in the code base

## Acceptance Criteria 
- [x] #5491 (?)
- [x] #5476 (3)
- [x] #5477 (1)
- [x] #5484 (2)
- [x] #5425 (5)
- [x] #5485 (5)

## Links 
- Related Epic (mandatory): #5312 

### in directory with existing devfile

``` 
Using devfile.yaml from the current directory.

Starting your application on cluster in developer mode ...
⠏ Waiting for Kubernetes resources ... DONE
⠏ Syncing files into the container ... DONE
⠏ Building your application in container on cluster ... DONE
⠏ Execting the application ... DONE
Your application is running on cluster.

You can access it at http://localhost:3000 and http://localhost:8080

⠏ Watching for changes in the current directory ... 

Press Ctrl+c to exit and clean up resources from cluster.

Change in main.java detected.
⠏ Syncing files into the container ... DONE
⠏ Reloading application ... DONE
⠏ Watching for changes in the current directory ...

Press Ctrl+c to exit and clean up resources from cluster.

<ctrl+c>

⠏ Cleaning up ... DONE                                                                             

```

### in a directory without devfile

```
$ odo dev
There is no devfile.yaml in the current directory.

Based on the files in the current directory odo detected
Language: Java
Project type: SpringBoot

? Is this correct? Yes

Current component configuration:
Opened ports:
- 8080
- 8084
Environment variables:
- FOO = bar
- FOO1 = bar1

? What configuration do you want change?  [Use arrows to move, type to filter]
> NOTHING - configuration is correct
  Delete port "8080"
  Delete port "8084"
  Add new port
  Delete environment variable "FOO"
  Delete environment variable "FOO1"
  Add new environment variable


⠏ Downloading "java-quarkus". DONE
Starting your application on cluster in developer mode ...
⠏ Waiting for Kubernetes resources ... DONE
⠏ Syncing files into the container ... DONE
⠏ Building your application in container on cluster ... DONE
⠏ Execting the application ... DONE
Your application is running on cluster.

You can access it at http://localhost:3000 and http://localhost:8080

⠏ Watching for changes in the current directory ... 

Press Ctrl+c to exit and clean up resources from cluster.

Change in main.java detected.
⠏ Syncing files into the container ... DONE
⠏ Reloading application ... DONE
⠏ Watching for changes in the current directory ...

Press Ctrl+c to exit and clean up resources from cluster.

<ctrl+c>

⠏ Cleaning up ... DONE
```

**Comment by dharmit** [comment]
WIP

* Do we need adapters in the code anymore? I admit to never having understood them well in the first place.
* What run modes to focus on? Only Run? Both Run and Debug?
* Code for `odo watch` can be re-used, but needs to be trimmed to remove cases/flags we don't want to cover.
* Port forwarding instead of using Ingress/Routes.

I'll start creating a "design document" as a part of the work on this issue. It will be a living document where we explain in plain English what a command would do in various scenarios.

**Comment by dharmit** [comment]
What's done so far (#5441):
* `odo dev` pushes to the Kubernetes cluster and watches for code changes.

What's remaining:
* Integration tests 
* Documentation
* Cleanup some hard coded items from the draft PR.

**Comment by dharmit** [comment]
All issues/stories in acceptance criteria are addressed.

/close
